protostar/cheatable_starknet/controllers/expect_events_controller.py

```diff
@@ -0,0 +1,166 @@
+
+
+@dataclass
+class EventMatching:
+    pass
+
+
+@dataclass
+class SkippedEventMatching(EventMatching):
+    emitted_event: Event
+
+
+@dataclass
+class AcceptedEventMatching(EventMatching):
+    emitted_event: Event
+    expected_event: Event
+
+
+@dataclass
+class FailedEventMatching(EventMatching):
+    expected_event: Event
+
+
+@dataclass
+class EventMatchingResult:
+    event_matchings: list[EventMatching]
+
+    @property
+    def failed_event_matchings(self):
+        return [
+            matching
+            for matching in self.event_matchings
+            if isinstance(matching, FailedEventMatching)
+        ]
+
+    @property
+    def should_be_accepted(self) -> bool:
+        return not self.failed_event_matchings
+
+
+class ExpectEventsController:
+    def __init__(
+        self,
+        test_finish_hook: "Hook",
+        test_execution_state: "CairoTestExecutionState",
+        cheatable_state: "CheatableCachedState",
+    ) -> None:
+        self._test_execution_state = test_execution_state
+        self._test_finish_hook = test_finish_hook
+        self._cheatable_state = cheatable_state
+        self._test_finish_hook.on(self.compare_expected_and_actual_results)
+
+    def add_expected_events(self, expected_events: list[Event]):
+        self._test_execution_state.expected_events_list.append(expected_events)
+
+    def compare_expected_and_actual_results(self):
+        for expected_events in self._test_execution_state.expected_events_list:
+            matching_result = match_events(
+                expected_events=expected_events,
+                emitted_events=self._cheatable_state.emitted_events,
+            )
+            if not matching_result.should_be_accepted:
+                raise ExpectEventsMismatchReportedException(matching_result)
+
+
+def match_events(
+    expected_events: list[Event], emitted_events: list[Event]
+) -> EventMatchingResult:
+    event_matchings: list[EventMatching] = []
+    for emitted_event in emitted_events:
+        if not expected_events or not should_accept_event_matching(
+            expected_event=expected_events[0], emitted_event=emitted_event
+        ):
+            event_matchings.append(SkippedEventMatching(emitted_event=emitted_event))
+        else:
+            event_matchings.append(
+                AcceptedEventMatching(
+                    emitted_event=emitted_event,
+                    expected_event=expected_events[0],
+                )
+            )
+            expected_events.pop(0)
+    failed_event_matchings = [
+        FailedEventMatching(expected_event) for expected_event in expected_events
+    ]
+    return EventMatchingResult(
+        event_matchings=[*event_matchings, *failed_event_matchings],
+    )
+
+
+def should_accept_event_matching(expected_event: Event, emitted_event: Event) -> bool:
+    return (
+        expected_event.key == emitted_event.key
+        and (expected_event.data is None or (expected_event.data == emitted_event.data))
```

Comment:
Hmm so if we expect no data but receive some data then we just ignore this data and consider this correct, right? It feels weird, afaik, for the expected call we went for the strict match, why wouldn't we do the same here?

Reply:
If data is not specified, then data is not compared. However, a user can check if there's no data was emitted by providing an empty list. This approach gives the user a choice.

Reply:
I understand, it feels like you basically explained what we both know. I still don't understand why we don't go for a "strict" match here (so if the emitted data is different than expected, it will be a fail for sure) and instead we use the "loose" match (if the expected emitted data is different than the actual emitted data, it's not always a fail).

Reply:
To give a user a choice. If he wants to compare the data, he will provide the data.

@Arcticae WDYT

Reply:
Seems reasonable to me to have no expectations from the data if the user doesn't

Reply:
Ok, resolving then.